### PR TITLE
FEATURE: Add ``TYPO3.Neos:ContentElementEditable`` TypoScript-Object

### PR DESCRIPTION
**Pull request by mficzel**
This change extracts the functionality of the neos:contentElement.editable viewHelper into a service and adds an equivalent TypoScript-Prototype.

TypoScript Example:
```
element = TYPO3.TypoScript:Array {
	title = TYPO3.TypoScript:Tag {
		tagName = 'h1'
		content = ${q(node).property('title')}
		@process.contentElementEditable = TYPO3.Neos:ContentElementEditable
		@process.contentElementEditable.property = 'title'
	}
	@process.contentElementWrapping = TYPO3.Neos:ContentElementWrapping
}
```

### CODE REVIEW

**Comment by mficzel** [comment]
reopen to run tests again

**Comment by mficzel** [comment]
I will check the broken tests the next days. Did'nt find them beforehand.

**Comment by johannessteu** [comment]
BTW @mficzel you can trigger the tests on the travis page. @kdambekalns showed me that trick some time ago :))

File: TYPO3.Neos/Classes/TYPO3/Neos/TypoScript/ContentElementEditableImplementation.php [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    public function getValue()
+    {
+        return $this->tsValue('value');
+    }
+
+    /**
+     * Evaluate this TypoScript object and return the result
+     *
+     * @return mixed
+     * @throws \TYPO3\Neos\Domain\Exception
```
Comment (aertmann) [inline TYPO3.Neos/Classes/TYPO3/Neos/TypoScript/ContentElementEditableImplementation.php]: not quite

**Comment by kitsunet** [comment]
Basically all the failing tests must be changed to test the service instead of the viewhelper now.

**Comment by aertmann** [comment]
nice change.. tests are broken though so they need to be fixed.. they re-run automatically when the code is updated

File: TYPO3.Neos/Classes/TYPO3/Neos/TypoScript/ContentElementEditableImplementation.php (hunk 1 above)
Comment (mficzel) [inline TYPO3.Neos/Classes/TYPO3/Neos/TypoScript/ContentElementEditableImplementation.php]: That exception is analog to the ContentElementWrapping ... maybe it is obsolete there aswell.

Comment (aertmann) [inline TYPO3.Neos/Classes/TYPO3/Neos/TypoScript/ContentElementEditableImplementation.php]: doesn't really matter, it's not thrown in this method and only added when using IDEs and hard to maintain and barely useful.. thus we usually avoid nested exception annotations

Comment (aertmann) [inline TYPO3.Neos/Classes/TYPO3/Neos/TypoScript/ContentElementEditableImplementation.php]: but it should also be removed in https://github.com/neos/neos-development-collection/blob/master/TYPO3.Neos/Classes/TYPO3/Neos/TypoScript/ContentElementWrappingImplementation.php#L52

**Comment by mficzel** [comment]
\o/ the tests are working now.

**Comment by dimaip** [comment]
:+1: by reading.

Crying from joy looking at this PR!

**Comment by johannessteu** [comment]
Yeah, totally! Maybe we could also alter our nodetypes like headline, text and so on to just wrap the contentelement stuff in backend mode so we get rid of the extra div in the frontend!

**Comment by kitsunet** [comment]
I am still against that because it means we deliver different markup (in terms of nesting) in frontend and backend and I thing that's a bad idea and can lead to hard to solve styling issues. IMHO it should be a concious decision of the implementer to make the markup diverge.

**Comment by bwaidelich** [comment]
> I am still against that because it means we deliver different markup [...] in frontend and backend

I agree!
With the rise of component-based markups (i.e. Web Components, ReactJS, ...) I would expect each node to be encapsulated in a single tag anyways

**Comment by johannessteu** [comment]
Hey @bwaidelich @kitsunet ,
i totally agree from that point of view. Especially about the component stuff. But i think we should still come up with a solution for markup like this
screen shot 2016-05-20 at 18 16 29

If you go through neos refrences you will find something like this everywhere. This is a neos.io screenshot btw

**Comment by bwaidelich** [comment]
True
```html
<div data-nodetype="typo3-neos-contentcollection">
  <h1 data-nodetype="typo3-neos-nodetypes-headline">..</h1>
</div>
```

should be everything that's needed here.. or even sth like
```html
<content-collection>
  <headline>..</headline>
</content-collection>
````

sometimes in the future.. semantic ftw

**Comment by dimaip** [comment]
@johannessteu @bwaidelich but you can already do that, specifying the `tagName`, no?

**Comment by johannessteu** [comment]
@dimaip depends. In general you can. But the `TYPO3.Neos:NodeTypes` e.g. wraps a `attributes` ts Object. And you can't pass that to the editable VH

**Comment by dimaip** [comment]
I mean you can specify `tagName` and I think even clasname for editable VH

**Comment by johannessteu** [comment]
Yes, class is possible. But you can't use the `TYPO3.TypoScript:Attributes` object we provide within the Content prototype as it renders also the `class=""` part and much more. So there are always two HTML-Tags for that case

```
<div {attributes -> f:format.raw()}">
    {neos:contentElement.editable(property: 'title', class:'{class}')}
</div>
```

**Comment by mficzel** [comment]
This discussion is not directly related to the PR or did I miss something here.

**Comment by aertmann** [comment]
true, let's avoid side-tracking this PR :) 👍 from reading

**Pull request merged**